security/postmortems/cve-2019-15225.md

```diff
@@ -0,0 +1,138 @@
+
+The fix re-implemented the `HeaderMapImpl::byteSize()` method to have O(1) performance by returning a `cached_byte_size_` member to `HeaderMapImpl` that was updated as header entries are added, rather than iterate over the `HeaderMap` to calculate the byte size. To resolve excessive iterations over the `HeaderMap` that can appear in access logging with many header formatters and many headers, the fix also included configurable limits for the maximum number of headers. 
+
+The following patches were produced:
+https://github.com/envoyproxy/envoy/commit/afc39bea36fd436e54262f150c009e8d72db5014
+https://github.com/envoyproxy/envoy/commit/5c122a35ebd7d3f7678b0f1c9846c1e282bba079
+
+A 1.11.2 security release was announced on 09-18-2019. An e-mail was sent to the Envoy private distributor list sharing the details of CVE-2019-15226. A week later, the candidate fix patches for CVE-2019-15226 were shared with distributors on 2019-09-24. This provided two weeks for distributors to test and prepare their software for the security release date, as per the guidelines set in place after security release 1.9.1.
+
+## Detection
+
+CVE-2019-15225 was reported by Seikun Kambashi in a public GitHub Issue describing a crash caused by a request with a very large URI for routes configured with a regex matcher: https://github.com/envoyproxy/envoy/issues/7728. 
+
+Envoy’s `route_fuzz_test`, which fuzzes route resolution and header finalization, ideally should have caught this crash. The test takes a `RouteConfiguration` and a set of headers as inputs, and routes a request with the input headers with the `RouteConfiguration` given. It should have been fairly easy for the fuzzers to produce a wildcard matcher and a long header string. However, the fuzz test itself had a logical error that resulted in ignoring input path headers and setting them to a default value of “/”. As a result, the fuzz test would never have tested a large URI and an OOM or crash would never have been detected. The fuzz test was fixed in https://github.com/envoyproxy/envoy/pull/8653, and a reproducer for the CVE was added.
+
+The underlying issue behind CVE-2019-15226 was first noticed via fuzzers when a timeout was reported by `h1_capture_direct_fuzz_test`: https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=16325. Some profiling work revealed that `HeaderMapImpl::byteSize()`, which is O(n) in the number of headers, is called for every single header in both HTTP/1.1 and HTTP/2 codecs. Although Envoy’s stateless HTTP/2 header fuzzers (`request_header_fuzz_test` and `response_header_fuzz_test`) perform 10x more executions per second than this fuzzer, these tested one header frame per testcase and used nghttp2’s default max header frame size (16 KB). Because of this, the frame size was too small to amplify the effect of the O(n^2) process enough to produce a timeout.
+
+## Action Items
+
+* https://github.com/envoyproxy/envoy/issues/8567 
```

Comment:
We should have some more followup issues created as per things that went well or we got lucky, ideally everything identified gets turned into a concrete AI.